### ISSUE TITLE
Support virtio-rng

### ISSUE THREAD

**Issue by nmeyerhans** [firecracker-microvm/firecracker#1671]
Per #663, the current recommendation for Firecracker is that guests rely on RDRAND (or similar) to seed the entropy pool used by the kernel's PRNG, enabling it with `random.trust_cpu` where necessary.  However, not all CPU architectures provide a facility for the generation and retrieval of random data. On these architectures, guests are left without a source of initial entropy, and must resort to jitter sampling and other less-than-ideal techniques. The effect is that, on architectures such as aarch64, guests either take longer to boot while applications block on data from `/dev/random` or `getrandom(2)`, or risk generating dangerously weak private key material if obtaining random data from `/dev/urandom`.

[virtio-rng](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/char/hw_random/virtio-rng.c) provides a mechanism to pass random data from the host to the guest for the purposes of seeding the initial entropy pool. Providing virtio-rng support would provide Firecracker guests with a uniform, architecture-independent, mechanism for initialization of the kernel's entropy pool at boot.

**Comment by bbros-dev** [comment]
Please see the [comment](https://github.com/firecracker-microvm/firecracker/issues/663#issuecomment-599107360) to issue #663

**Comment by petreeftime** [comment]
Virtio-rng is relatively simple to build, but some issues need to be clarified:
* Where is the entropy coming from? Is /dev/urandom  or the rand crate good enough? Should /dev/random be used instead?
* How do you make sure the host doesn't run out of entropy, especially if using /dev/random? This might be relevant only when running many VMs. You would need some throttling in place either way.

**Comment by nmeyerhans** [comment]
> Virtio-rng is relatively simple to build, but some issues need to be clarified:
> 
>     * Where is the entropy coming from? Is /dev/urandom  or the rand crate good enough? Should /dev/random be used instead?
> 
>     * How do you make sure the host doesn't run out of entropy, especially if using /dev/random? This might be relevant only when running many VMs. You would need some throttling in place either way.

I'll suggest that you probably want to use [getrandom(2)](http://man7.org/linux/man-pages/man2/getrandom.2.html) rather than device access. Among other things, it improves the behavior when accessing the "urandom" datasource in that it blocks if the CSPRNG has not yet been initialized, in that it blocks in that case. Otherwise, similar to accessing `/dev/urandom`, it is nonblocking.

Thomas Hühn has written a decent [analysis](https://www.2uo.de/myths-about-urandom/) of the Linux CSPRNG subsystem comparing the "random" and "urandom" sources.

**Comment by bbros-dev** [comment]
My understanding is the kernel authors (finally?) are likely to remove the kernel from the random numbers game.

Regardless of whether or not that comes to pass, I'd urge that any `virtio-rng` be limited to giving a guest VM access to any hardware RNG device(s) on the host.
Specifically, the `virtio-rng`  should not expose any host software level randomness.

As a workaround for the evolving state of play it is possible for a user to:

1. pass in a random sequence 
1. deplete the entropy
1. Insert the random data
1. update the entropy count

Right now that is cumbersome and code driven.

Perhaps Firecracker could, when no host RNG hardware is available, provide an interface so that the `virtio-rng` delivers the user provided 'one shot' random sequence to the guest.  

Further additions to the pool of random numbers after the one shot initialization are the users responsibility via `haveged` or some such package.

In this scenario it actually would help if Firecracker could deliver the host without any entropy at the point immediately before the user provided one-shot random data is inserted into `virt-io` (or before host hardware provided RNG is connected).

Hope that makes sense?

**Comment by clintonPantheon** [comment]
Hi, 

I'm new here.  I've been looking at firecracker as a possible solution to a problem at work but it seems that rng is still an issue here for virtualization.  

### CPU Time Execution Jitter for Virtualization Problems

@smuellerDD has spent quite some time incorporating CPU time execution jitter as an entropy source for the [kernel](https://elixir.bootlin.com/linux/v5.8-rc4/source/crypto/jitterentropy.c). One of the use cases it attempts to address is [entropy for virtualization](https://www.chronox.de/jent/) to support cryptographic applications (e.g. openssl, sshd, etc.). 

There is a jitterentropy daemon that  might help here to seed entropy early on in the startup process of a new microvm.  

https://github.com/smuellerDD/jitterentropy-rngd

One problem I see in this issue is:
>not all CPU architectures provide a facility for the generation and retrieval of random data

and in #663, the issue seems to be an exhaustion of entropy. 

Is there a way to support the the firecracker KVM to utilize entropy from:
1.  `jitterentropy-rngd` or,
2. have the kvm support `virtio-rng` and have `jitterentropy-rngd` as an entropy source for the`virtio-rng`?

**Comment by petreeftime** [comment]
> Regardless of whether or not that comes to pass, I'd urge that any `virtio-rng` be limited to giving a guest VM access to any hardware RNG device(s) on the host.
> Specifically, the `virtio-rng` should not expose any host software level randomness.

I disagree with the premise that software level entropy is bad, especially if it's a mixture of multiple hardware level sources. What exactly is the threat you're trying to defend against with only hardware level interfaces? The VM itself has no visibility of where the entropy comes from anyway.

**Comment by bbros-dev** [comment]
> I disagree with the premise that software level entropy is bad....

OK. We're not assessing that topic. For others following this issue there is some context in the documentation issue #663 

> especially if it's a mixture of multiple hardware level sources.

Especially then it wouldn't be software level.

> What exactly is the threat you're trying to defend against with only hardware level interfaces?

I don't believe I ever suggested hardware only, just the opposite I believe.
The 'threat' is real and occurs now with VM's freezing waiting for entropy levels to rise.

> The VM itself has no visibility of where the entropy comes from anyway.

True, but this issue and #663 concern the VM owner/operator, not the VM.  My concern is the presence of a virtio-rng device creates the impression there is a secure/reliable rng available.  I've pointed out that it appears the kernel is possibly going to move out of this business - IMHO that is a good thing.  I've simply argued that if the host kernel is going to provide insecure/non-unique numbers then a reasonable approach is for the virtio-rng to passthrough a) numbers from a hardware device b) entropy values 'user-space' provides - I listed the steps a user currently has to do in their own code - the virtio-rng could provide that functionality.

I don't dispute that the user may install somethhing like haveged and a shim to feed the virtio-rng.  Sorry if that wasn't obvious.
